crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs

```diff
@@ -0,0 +1,386 @@
+};
+use biome_rowan::{AstNode, AstNodeList, AstSeparatedList, TextRange};
+use biome_rule_options::no_setup_props_reactivity_loss::NoSetupPropsReactivityLossOptions;
+
+declare_lint_rule! {
+    /// Disallow usages that lose the reactivity of `props` passed to `setup`.
+    ///
+    /// Vue's Composition API requires that props passed to the `setup` function
+    /// maintain reactivity. Destructuring props or using member expressions on props
+    /// in the root scope of `setup` will cause the values to lose reactivity.
+    ///
+    /// This rule reports:
+    /// - Direct destructuring of props in setup function parameters
+    /// - Destructuring assignment of props in the root scope of setup
+    /// - Member access patterns that could break reactivity
+    ///
+    /// Note: Destructuring is allowed inside nested functions, callbacks, and
+    /// returned render functions where the reactive context is preserved.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// // Destructuring in setup parameters
+    /// export default {
+    ///   setup({ count }) {
+    ///     // count is no longer reactive
+    ///     return () => h('div', count)
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// // Destructuring in setup root scope
+    /// export default {
+    ///   setup(props) {
+    ///     const { count } = props
+    ///     // count is no longer reactive
+    ///     return () => h('div', count)
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// // Keep props reactive
+    /// export default {
+    ///   setup(props) {
+    ///     watch(() => props.count, () => {
+    ///       console.log(props.count)
+    ///     })
+    ///     return () => h('div', props.count)
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// ```js
+    /// // Destructuring inside callbacks is OK
+    /// export default {
+    ///   setup(props) {
+    ///     watch(() => props.count, () => {
+    ///       const { count } = props // OK inside callback
+    ///       console.log(count)
+    ///     })
+    ///     return () => {
+    ///       const { count } = props // OK inside render function
+    ///       return h('div', count)
+    ///     }
+    ///   }
+    /// }
+    /// ```
+    ///
+    pub NoSetupPropsReactivityLoss {
+        version: "next",
+        name: "noSetupPropsReactivityLoss",
+        language: "js",
+        recommended: false,
+    }
+}
+
+/// Represents the different types of reactivity loss violations that can occur.
+pub enum RuleState {
+    /// Props destructured directly in setup function parameters.
+    DestructuredParameter(TextRange),
+    /// Props destructured in variable assignments within the setup function body.
+    DestructuredAssignment(TextRange),
+}
+
+impl Rule for NoSetupPropsReactivityLoss {
+    type Query = Semantic<JsExportDefaultExpressionClause>;
```

Comment:
This rule isn't using the semantic model. Replace `Semantic` with the `Ast` query